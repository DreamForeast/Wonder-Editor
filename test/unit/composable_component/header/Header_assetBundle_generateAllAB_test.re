open Wonder_jest;

open Expect;

open Expect.Operators;

open Sinon;

open Js.Promise;

open WonderBsMost;

let _ =
  describe("Header AssetBundle->generate all ab", () => {
    let sandbox = getSandboxDefaultVal();

    beforeEach(() => {
      sandbox := createSandbox();

      MainEditorSceneTool.initState(~sandbox, ());

      MainEditorSceneTool.prepareScene(sandbox);

      MainEditorAssetTool.buildFakeFileReader();
    });
    afterEach(() => restoreSandbox(refJsObjToSandbox(sandbox^)));

    describe("generate all ab", () => {
      describe("test buildSelectTreeForGenerateAllAB", () =>
        testPromise("should only has asset bundle assets", () => {
          let addedMaterialNodeId1 = MainEditorAssetIdTool.getNewAssetId();
          MainEditorAssetHeaderOperateNodeTool.addMaterial();

          let addedFolderNodeId1 = MainEditorAssetIdTool.getNewAssetId();
          MainEditorAssetHeaderOperateNodeTool.addFolder();

          MainEditorAssetTreeTool.Select.selectFolderNode(
            ~nodeId=addedFolderNodeId1,
            (),
          );

          MainEditorAssetUploadTool.loadOneTexture(
            ~imgName="image1.png",
            ~imgSrc="newImgBase64",
            (),
          )
          |> then_(uploadedTextureNodeId =>
               MainEditorAssetUploadTool.loadOneAssetBundle(
                 ~fileName="A.rab",
                 (),
               )
               |> then_(uploadedAssetBundleNodeId =>
                    HeaderAssetBundleTool.GenerateAllAB.buildGenerateAllABModal(
                      ~selectTree=
                        HeaderAssetBundleTool.GenerateAllAB.buildSelectTreeForGenerateAllAB
                        |> StateLogicService.getStateToGetData,
                      ~send=SinonTool.createOneLengthStub(sandbox^),
                      (),
                    )
                    |> BuildComponentTool.buildUI
                    |> ReactTestTool.createSnapshotAndMatch
                    |> resolve
                  )
             );
        })
      );

      describe("test generate all ab from selected single abs", () => {
        let _getZipFilePath = (file, callCount) =>
          file
          |> getCall(callCount)
          |> getArgs
          |> Js.List.hd
          |> OptionService.unsafeGet;

        beforeEach(() => {
          LoadTool.buildFakeTextDecoder(LoadTool.convertUint8ArrayToBuffer);
          LoadTool.buildFakeTextEncoder();
          LoadTool.buildFakeURL(sandbox^);

          LoadTool.buildFakeLoadImage(.);

          HeaderAssetBundleTool.GenerateAllAB.prepareDigest(sandbox^);
        });

        describe("test generated abs->relative path", () =>
          testPromise("generated abs->relative path === the path in zip", () => {
            MainEditorSceneTool.createDefaultSceneAndNotInit(sandbox);

            let addedMaterialNodeId1 = MainEditorAssetIdTool.getNewAssetId();
            MainEditorAssetHeaderOperateNodeTool.addMaterial();

            let selectTree =
              HeaderAssetBundleTool.GenerateSingleRAB.buildSelectTreeForGenerateSingleRAB
              |> StateLogicService.getStateToGetData
              |> SelectTreeTool.setSelectForSelectTree(
                   true,
                   MainEditorAssetMaterialNodeTool.getMaterialName(
                     ~nodeId=addedMaterialNodeId1,
                     (),
                   ),
                 );

            let rab =
              HeaderAssetBundleTool.GenerateSingleRAB.generateSingleRAB(
                ~selectTree,
                (),
              );

            let sab =
              HeaderAssetBundleTool.GenerateSingleSAB.generateSingleSAB();

            MainEditorAssetUploadTool.loadOneAssetBundle(
              ~fileName="A.rab",
              ~assetBundle=rab,
              (),
            )
            |> then_(uploadedRABAssetBundleNodeId => {
                 let addedFolderNodeId1 =
                   MainEditorAssetIdTool.getNewAssetId();
                 MainEditorAssetHeaderOperateNodeTool.addFolder();

                 MainEditorAssetTreeTool.Select.selectFolderNode(
                   ~nodeId=addedFolderNodeId1,
                   (),
                 );

                 MainEditorAssetUploadTool.loadOneAssetBundle(
                   ~fileName="B.sab",
                   ~assetBundle=sab,
                   (),
                 )
                 |> then_(uploadedSABAssetBundleNodeId => {
                      let obj =
                        HeaderTool.buildPublishFakeJsZipCreateFunc(sandbox^);

                      let selectTree =
                        HeaderAssetBundleTool.GenerateAllAB.buildSelectTreeForGenerateAllAB
                        |> StateLogicService.getStateToGetData
                        |> SelectTreeTool.setSelectForSelectTree(
                             true,
                             MainEditorAssetAssetBundleNodeTool.getName(
                               uploadedRABAssetBundleNodeId,
                             )
                             |> StateLogicService.getEditorState,
                           )
                        |> SelectTreeTool.setSelectForSelectTree(
                             true,
                             MainEditorAssetAssetBundleNodeTool.getName(
                               uploadedSABAssetBundleNodeId,
                             )
                             |> StateLogicService.getEditorState,
                           );

                      HeaderAssetBundleTool.GenerateAllAB.generateAllABZip(
                        ~selectTree,
                        ~dependencyRelationInputValue=
                          HeaderAssetBundleGenerateAllAB.Method.buildDefaultDependencyRelationInputValue(),
                        ~createZipFunc=() => obj,
                        (),
                      )
                      |> Most.drain
                      |> then_(() => {
                           let file = obj##file;

                           (
                             file |> getCallCount,
                             _getZipFilePath(file, 0),
                             _getZipFilePath(file, 1),
                             _getZipFilePath(file, 2),
                           )
                           |> expect
                           == (
                                3,
                                "A.rab",
                                "New Folder/B.sab",
                                "WonderWAB.wab",
                              )
                           |> resolve;
                         });
                    });
               });
          })
        );

        testPromise("error should be catched by stream", () => {
          let errorMsg = ref("");

          MainEditorAssetUploadTool.loadOneAssetBundle()
          |> then_(uploadedSABAssetBundleNodeId => {
               let obj = HeaderTool.buildPublishFakeJsZipCreateFunc(sandbox^);

               let selectTree =
                 HeaderAssetBundleTool.GenerateAllAB.buildSelectTreeForGenerateAllAB
                 |> StateLogicService.getStateToGetData
                 |> SelectTreeTool.setSelectForSelectTree(
                      true,
                      MainEditorAssetAssetBundleNodeTool.getName(
                        uploadedSABAssetBundleNodeId,
                      )
                      |> StateLogicService.getEditorState,
                    );

               HeaderAssetBundleTool.GenerateAllAB.generateAllABZip(
                 ~selectTree,
                 ~dependencyRelationInputValue={|
    ttt
    |},
                 ~createZipFunc=() => obj,
                 (),
               )
               |> Most.drain
               |> catch(e => {
                    let message: string = Obj.magic(e)##message;

                    errorMsg := message;

                    () |> Obj.magic;
                  })
               |> then_(() =>
                    errorMsg^ |> expect == "ttt is not defined" |> resolve
                  );
             });
        });
      });
    });
  });