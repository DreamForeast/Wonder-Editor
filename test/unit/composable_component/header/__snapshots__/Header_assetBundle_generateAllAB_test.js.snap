// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header AssetBundle->generate all ab generate all ab test buildSelectTreeForGenerateAllAB should only has asset bundle assets 1`] = `
<article
  className="wonder-generateAllAB-modal"
>
  <div
    className="modal-item"
  >
    <div
      className="modal-item-header"
    >
      Generate All AB-&gt;zip from multiple 'Single AB'
      <img
        onClick={[Function]}
        src="./public/img/close.png"
      />
    </div>
    <div
      className="modal-item-content"
    >
      <div
        className="content-field"
      >
        <div
          className="field-title"
        >
          name
        </div>
        <div
          className="field-content"
        >
          <input
            className="input-component"
            onChange={[Function]}
            type="text"
            value="WonderAllAB"
          />
        </div>
      </div>
      <article
        className="wonder-textAreaInput"
      >
        <div
          className="content-field content-textarea"
        >
          <div
            className="field-title"
          >
            relation
          </div>
          <div
            className="field-content"
          >
            <textarea
              className="input-component"
              onChange={[Function]}
              type="text"
              value="
    (function() {
        var dependencyRelation = {};

        //todo need rewrite
        dependencyRelation[\\"A.sab\\"] = [\\"AssetBundle/B.rab\\", \\"AssetBundle/F/c.rab\\"];

        return dependencyRelation;
    }())
    "
            />
          </div>
        </div>
      </article>
      <article
        className="wonder-selectTree"
      >
        <ul
          className="wonder-tree-node"
        >
          <li>
            <div
              className="tree-container"
            >
              <input
                defaultChecked={false}
                onClick={[Function]}
                type="checkbox"
              />
              <img
                className="treeNode-icon"
                src="./public/img/selectFolder.png"
              />
              Assets
            </div>
          </li>
          <ul
            className="wonder-tree-node"
          >
            <li>
              <div
                className="tree-container"
              >
                <input
                  defaultChecked={false}
                  onClick={[Function]}
                  type="checkbox"
                />
                <img
                  className="treeNode-icon"
                  src="./public/img/selectFolder.png"
                />
                New Folder
              </div>
            </li>
            <ul
              className="wonder-tree-node"
            >
              <li>
                <div
                  className="tree-container"
                >
                  <input
                    defaultChecked={false}
                    onClick={[Function]}
                    type="checkbox"
                  />
                  <img
                    className="treeNode-icon"
                    src="./public/img/assetBundle.png"
                  />
                  A
                </div>
              </li>
            </ul>
          </ul>
        </ul>
      </article>
    </div>
    <div
      className="modal-item-footer"
    >
      <button
        className="footer-submit"
        onClick={[Function]}
      >
        Submit
      </button>
    </div>
  </div>
</article>
`;
